### PR TITLE
feat(op_crates/web): Adding uft-16 and big5

### PR DESCRIPTION
**Pull request by AliBasicCoder**
This PR adds encoding utf-16 and big5
NOTE: need more testing

### CODE REVIEW

**Review by crypt096 — APPROVED** [review]

**Review by bartlomieju — DISMISSED** [review]
Marking as "Request changes" because of `NOTE: need more testing`

**Comment by AliBasicCoder** [comment]
> Marking as "Request changes" because of `NOTE: need more testing`

do you know where to find tests??

**Comment by AliBasicCoder** [comment]
> Marking as "Request changes" because of `NOTE: need more testing`

should big5 char points array written in a single line or 10 (or more) items per line??
NOTE: splitting big5 char points into 10 per line would add 1979 lines

**Comment by AliBasicCoder** [comment]
@bartlomieju big5 is now tested but i can't find any testing for utf-16 in [wpt](https://github.com/web-platform-tests/wpt/) except already added and i don't think that's enough testing

**Comment by AliBasicCoder** [comment]
HUGE NOTE: because of tests array in big5Test function in op_crates/web/text_encoding_test.js editing becomes really hard in vscode especially with `Bracket Pair Colorizer` extension enabled

**Comment by kitsonk** [comment]
@AliBasicCoder is there anyway to create them as external fixtures in the file system?  @bartlomieju seems like for complex stuff like this for `op_crates` we need something that allows the ability to read fixtures from the file system.

**Comment by AliBasicCoder** [comment]
should i move them

**Comment by AliBasicCoder** [comment]
@kitsonk should i move them??

**Comment by bartlomieju** [comment]
@AliBasicCoder there's no way to load fixtures from files right now, let's keep them where they are right now.

**Comment by AliBasicCoder** [comment]
> @AliBasicCoder there's no way to load fixtures from files right now, let's keep them where they are right now.

i recommend moving them in another js file so editing 08_text_encoding.js stay easy

**Comment by bartlomieju** [comment]
> > @AliBasicCoder there's no way to load fixtures from files right now, let's keep them where they are right now.
> 
> i recommend moving them in another js file so editing 08_text_encoding.js stay easy

Sounds good to me

**Comment by AliBasicCoder** [comment]
@bartlomieju utf-16 is now tested i and i think this PR is ready to merge

**Comment by linux-china** [comment]
@AliBasicCoder  thanks for your work.   AssemblyScript's as-bind uses "utf-16le" now.  

```
error: Uncaught RangeError: The encoding label provided ('utf-16le') is invalid.
```

**Comment by AliBasicCoder** [comment]
@bartlomieju @ry @kitsonk this PR is ready

**Comment by kitsonk** [comment]
Going to move this to 1.7.0 as it add quite a lot of "static" decoding tables to the runtime and we need more time to figure out the right path forward.

**Comment by lucacasonato** [comment]
@AliBasicCoder Can you rebase this? We now have WPT for testing. You can use this to test these encodings. To enable the `utf16` / `big5` tests, see `cli/tests/wpt.jsonc`.

File: op_crates/web/08_text_encoding.js [hunk 1]
```diff
@@ -138,6 +138,124 @@
...
+
+  function Utf16ByteDecoder(
+    bytes,
+    be = false,
+    fatal = false,
+    ignoreBOM = false,
+  ) {
+    if (ignoreBOM) {
+      throw new TypeError("Ignoring the BOM is available only with utf-8.");
+    }
```
Comment (lucacasonato) [inline op_crates/web/08_text_encoding.js:210]: UTF-16 has BOM too. It is `0xfe, 0xff` / `0xff, 0xfe`. See `wpt/encoding/textdecoder-byte-order-marks.any.js`.

**Comment by AliBasicCoder** [comment]
> @AliBasicCoder Can you rebase this? We now have WPT for testing. You can use this to test these encodings. To enable the `utf16` / `big5` tests, see `cli/tests/wpt.jsonc`.

should encoding tests in `op_crates` be removed

**Comment by AliBasicCoder** [comment]
> @AliBasicCoder Can you rebase this? We now have WPT for testing. You can use this to test these encodings. To enable the `utf16` / `big5` tests, see `cli/tests/wpt.jsonc`.

PR is rebased now and utf-16 & big5 tests is enabled

**Comment by AliBasicCoder** [comment]
an error that is not related to this PR happend

**Comment by lucacasonato** [comment]
Bartek and me came to the conclusion offline, that we can just use WPT instead of having our own tests for text encoding / decoding. I have updated the PR accordingly.

@bartlomieju PTAL

**Comment by lucacasonato** [comment]
@AliBasicCoder The tests for UTF-16 are not passing yet. Try `cargo test web_platform_tests`.

**Comment by AliBasicCoder** [comment]
> @AliBasicCoder The tests for UTF-16 are not passing yet. Try `cargo test web_platform_tests`.

it should be fixed now 
and for some reason github says there are conflicts and i don't have permission to fix them

**Comment by lucacasonato** [comment]
@AliBasicCoder No problem, I will resolve the conflicts.

**Review by lucacasonato — APPROVED** [review]
Thanks a bunch for implementing this. I am very sorry this has taken so long to land, and I hope this has not discouraged you to work on the TextEncoder implementation 😄. Now that we have WPT for all of this, we have a great idea of what bugs we still have to squash.

LGTM!

**Review by bartlomieju — APPROVED** [review]
LGTM

**Comment by linux-china** [comment]
@AliBasicCoder  It works well with Wasm + JS compiled from AssemblyScript and Emscripten.   Thanks.

**Comment by skanfd** [comment]
Will utf-16 be available in 1.7.0? :)

**Comment by grian32** [comment]
@skanfd looking at the [release notes for 1.7.0](https://github.com/denoland/deno/releases/tag/v1.7.0), yes.

**Comment by skanfd** [comment]
@grian32 Aha...I need this!

**Comment by skanfd** [comment]
@grian32 It does!
Thanks for all contributors~

**Pull request merged**